pkg/didcomm/protocol/introduce/service.go

```diff
@@ -127,31 +179,90 @@ func (s *Service) sendMsgEvents(msg service.StateMsg) {
 
 // sendEvent triggers the action event. This function stores the state of current processing and passes a callback
 // function in the event message.
-func (s *Service) sendActionEvent(msg *service.DIDCommMsg, aEvent chan<- service.DIDCommAction, nextState state) error {
+func (s *Service) sendActionEvent(msg *metaData, aEvent chan<- service.DIDCommAction) error {
+	// save the incoming message in the store (to retrieve later when callback events are fired)
+	ID := uuid.New().String()
+	err := s.save(ID, msg)
+	if err != nil {
+		return fmt.Errorf("JSON marshalling of document failed: %w", err)
+	}
+
+	// create the message for the channel
+	didCommAction := service.DIDCommAction{
+		ProtocolName: Introduce,
+		Message:      msg.Msg,
+		Continue: func() {
+			s.processCallback(ID, nil)
+		},
+		Stop: func(err error) {
+			s.processCallback(ID, err)
+		},
+		//Properties: createEventProperties(threadID, ""),
+	}
+
+	// trigger the registered action event
+	aEvent <- didCommAction
+
 	return nil
 }
 
-func nextState(msg *service.DIDCommMsg) (state, error) {
+func (s *Service) processCallback(id string, err error) {}
+
+func nextState(msg *service.DIDCommMsg, rec *record, outbound bool) (state, error) {
 	switch msg.Header.Type {
-	case ProposalMsgType:
+	// SkipProposal is an artificial message type, we need to replace it with a real one
+	case SkipProposalMsgType:
+		msg.Payload = []byte(strings.Replace(string(msg.Payload), SkipProposalMsgType, ProposalMsgType, 1))
+		rec.WaitCount--
 		return &arranging{}, nil
+	case ProposalMsgType:
+		if outbound {
+			return &arranging{}, nil
+		}
+		return &deciding{}, nil
 	case ResponseMsgType:
-		return &delivering{}, nil
+		if outbound {
+			return &waiting{}, nil
+		}
+		rec.WaitCount--
+		if rec.WaitCount == 0 {
+			return &delivering{}, nil
+		}
+		return &arranging{}, nil
+	case AckMsgType:
+		return &done{}, nil
 	default:
 		return nil, fmt.Errorf("unrecognized msgType: %s", msg.Header.Type)
 	}
 }
 
-func (s *Service) currentState(thID string) (state, error) {
-	name, err := s.store.Get(thID)
+func (s *Service) currentStateRecord(thID string) (*record, error) {
+	src, err := s.store.Get(thID)
+	if errors.Is(err, storage.ErrDataNotFound) {
+		return &record{
+			StateName: stateNameStart,
+			WaitCount: 2,
```

Comment:
is the value configurable ?

Reply:
probably in the future :) I found TODO in spec
`TODO:do we want to support introducing more than 2 at a time?`